vllm/model_executor/models/decilm.py

```diff
@@ -0,0 +1,119 @@
+                                              hf_model_weights_iterator)
+
+DECI_LM_MODEL_NAME = 'Deci/DeciLM-7b-instruct'
+
+
+class DeciLMForCausalLM(LlamaForCausalLM):
+    """
+    Implementation for https://huggingface.co/Deci/DeciLM-7b-instruct.
+    Based on the llama executor.
+
+    The main difference is that DeciLM uses Variable Grouped Query Attention.
+    The constant number of GQA heads in the decoder is overriden with a value per layer.
+
+    Usually, in the HuggingFace implementation, Instead of "config.num_key_value_heads",
+    We use "config.num_key_value_heads_per_layer[i]" which varies.
+
+    Currently, PagedAttention does not work well with variable GQA, so we normalize the
+    weights upon loading, and use uniform GQA with the max value instead.
+    """
+
+    def __init__(
+        self,
+        config: Optional[PretrainedConfig] = None,
+        linear_method: Optional[LinearMethodBase] = None,
+    ) -> None:
+        config.num_key_value_heads = max(config.num_key_value_heads_per_layer)
+        delattr(config, 'num_key_value_heads_per_layer')
+        super().__init__(config=config, linear_method=linear_method)
+
+    def load_weights(self,
+                     model_name_or_path: str,
+                     cache_dir: Optional[str] = None,
+                     load_format: str = "auto",
+                     revision: Optional[str] = None):
+        stacked_params_mapping = [
+            # (param_name, shard_name, shard_id)
+            ("qkv_proj", "q_proj", "q"),
+            ("qkv_proj", "k_proj", "k"),
+            ("qkv_proj", "v_proj", "v"),
+            ("gate_up_proj", "gate_proj", 0),
+            ("gate_up_proj", "up_proj", 1),
+        ]
+        params_dict = dict(self.named_parameters())
+        for name, loaded_weight in hf_model_weights_iterator(
+                model_name_or_path, cache_dir, load_format, revision):
+            if "rotary_emb.inv_freq" in name:
+                continue
+            if "rotary_emb.cos_cached" in name:
+                continue
+            if "rotary_emb.sin_cached" in name:
+                continue
+
+            if "k_proj" in name or "v_proj" in name:
+                loaded_weight = self.degroup_weight(loaded_weight)
+
+            for (param_name, weight_name, shard_id) in stacked_params_mapping:
+                if weight_name not in name:
+                    continue
+                param = params_dict[name.replace(weight_name, param_name)]
+                weight_loader = param.weight_loader
+                weight_loader(param, loaded_weight, shard_id)
+                break
+            else:
+                param = params_dict[name]
+                weight_loader = getattr(param, "weight_loader",
+                                        default_weight_loader)
+                weight_loader(param, loaded_weight)
+
+    def degroup_weight(self, loaded_weight: torch.Tensor) -> torch.Tensor:
```

Comment:
This method receives a weight and change the number of kv heads to meet the max. number that is found in the list: 
https://huggingface.co/Deci/DeciLM-7B/blob/main/config.json#L18

By doing so, all the attention layers will have the same number of heads, which is required for PagedAttention.

Reply:
Makes sense. Thanks for this solution. Let's leave it as future work for now.